### PR TITLE
[rush] implementation for defining PNPM's store path

### PR DESCRIPTION
**Pull request by Js-Brecht**
This PR implements a feature extending PNPM options to allow the end-user to define where they wish PNPM to put its store.  The conversation regarding the design can be found in issue https://github.com/microsoft/rushstack/issues/1414

Available options are under `pnpmOptions` in `rush.json`:

1. `local`: Use rush's method of storing in the common temp folder as `pnpm-store`.
2. `global`: Do no define a store path, allowing PNPM to determine itself where to put the store
    * By default, it stores it in the global store.
    * If the store path is defined in `.npmrc`, PNPM will use that
        * This is a side-effect.
3. `path`: Put store in path defined by `pnpmStorePath`.
    * `pnpmStorePath` is a path relative to the rush project root (where `rush.json` is found)
    * If `pnpmStorePath` is not defined, then the default (used by `local`) will be used
    * A new environment variable named `RUSH_PNPM_STORE_PATH` can be defined, and will be used if `pnpmStore = 'path'`.

To make sure that the store remains consistent between builds, and to keep pnpm from throwing errors in the CLI, I chose to persist the store path in `last-install.flag`.  I chose this method instead of using `node_modules/.modules.yaml` for the following reasons

1. There is no guarantee that `.modules.yaml` format will not change in the future
2. Finding the store that `pnpm` will use when `pnpmStore` is set to `global` is not readily accessible, and varies between versions of `pnpm`.
    * `pnpm` could also be determining its store location from other locations, including `.npmrc` or environment variables

_**Any time**_ that the desired store location differs from the last one used, it will throw an error on the command line suggesting that the user run `rush update --purge`.

---

* [x] assets
    * `rush.json` updated to include new options
* [x] schema
    * `pnpmOptions` updated to include `pnpmStore` and `pnpmStorePath`
* [x] EnvironmentConfiguration
    * Use `RUSH_PNPM_STORE_PATH` variable when it is present
* [x] RushConfiguration
    * Determine store location based on `pnpmStore` and `pnpmStorePath` + `RUSH_PNPM_STORE_PATH`
* [x] LastInstallFlag
    * Error handling, in the event `last-install.flag` state does not match new, desired, state
* [x] InstallManager
    * Use new `pnpmOptions.pnpmStorePath`
    * In the case `pnpmStore = 'global'`, `--store` flag will not be included in `pnpm`'s CLI flags
* [x] tests included to cover new functionality in the following modules
    * EnvironmentConfiguration
    * RushConfiguration
    * InstallManager
    * LastInstallFlag
* [x] manual testing in a test repository using the new feature, under the following circumstances:
    * Each test was performed with and without `--purge` flag, to ensure it aborts if the store path has changed, and also that it resets the store correctly.  Each test was also run with and without the environment variable present
        * `pnpmStore = 'local'`
        * `pnpmStore = 'global'`
        * `pnpmStore = 'path'`
             * With and without `pnpmStorePath`

---

Additional notes: 
* `rush change` didn't ask for an update type.  Not sure if that's a big deal or not.

### CODE REVIEW

**Comment by Js-Brecht** [comment]
Apparently there's a bug in [`EnvironmentConfiguration._normalizeDeepestParentFolderPath()`](https://github.com/Js-Brecht/rushstack/blob/pnpm-store-options/apps/rush-lib/src/api/EnvironmentConfiguration.ts#L204) when running on linux.  If the first folder following root doesn't exist, it will drop the preceding slash.  Not that that is particularly a bad thing in this instance, since this process shouldn't really be creating folders in the root directory (wasn't really thinking about that when I wrote the test, though).

I wonder if there should be some sort of fail-safe or check for that when running that function, though :thinking:.  It could lead to some unexpected behavior, if, for some odd reason, the user set `RUSH_PNPM_STORE_PATH` to `/nonexistant-folder`... same thing would apply if they set `RUSH_TEMP_FOLDER` to `/nonexistant-folder`, too.

That's beyond the scope of this PR, however, so I'm not going to worry about it.  I just fixed the test.

**Comment by octogonz** [comment]
> Apparently there's a bug in [`EnvironmentConfiguration._normalizeDeepestParentFolderPath()`](https://github.com/Js-Brecht/rushstack/blob/pnpm-store-options/apps/rush-lib/src/api/EnvironmentConfiguration.ts#L204) when running on linux.

Hrmm... yeah.  This was introduced by @iclanton in PR https://github.com/microsoft/rushstack/pull/1428.  I feel like the overall implementation could be better.

It seems that `true-case-path` starts from the root, and [loops over each path segment](https://github.com/Profiscience/true-case-path/blob/ac422f962cb449a574ebcfa62d9c58abad8ff5cb/index.js#L68) calling `readdir()` to determine the actual casing.  Whereas `_normalizeDeepestParentFolderPath()` essentially walks backwards doing the same thing.  That seems like an `O(n^2)` algorithm.  Seems like whatever problem this is solving ideally could be contributed directly to `true-case-path`, or else our version should reimplement its (relatively simple) algorithm.

@Js-Brecht could you open a separate GitHub issue for this?

**Comment by Js-Brecht** [comment]
> It seems that `true-case-path` starts from the root, and [loops over each path segment](https://github.com/Profiscience/true-case-path/blob/ac422f962cb449a574ebcfa62d9c58abad8ff5cb/index.js#L68)

Yup, and since the first folder doesn't exist, the entire path that is defined when that function is called is "/"; so on Linux, this results in the `segments` parameter being an empty array.  It never has the opportunity to [prepend the delimiter](https://github.com/Profiscience/true-case-path/blob/ac422f962cb449a574ebcfa62d9c58abad8ff5cb/index.js#L72).

> could you open a separate GitHub issue for this?

I can do that

**Comment by Js-Brecht** [comment]
@octogonz meant to ask before I hit accept.... would it be more appropriate to open the issue in this repository, or in https://github.com/Profiscience/true-case-path?

**Comment by octogonz** [comment]
That project looks actively maintained, so maybe we should propose to fix it there.

You can also open  Rush issue to track the work here, and we can tag it as "external"

**Comment by RIP21** [comment]
Can this be finally done before it gets full of merging conflicts, please, please, please? :) This will increase speeds of our CI significantly (local installs too tho).

**Comment by WhiteR4bbitt** [comment]
**tl;dr;** Would also love for this to get merged! ^

Over the last few days I have been looking for something to manage my *monorepo* project I am attempting and came across `pnpm` (after trying `lerna`..) which lead me to combine it with `rush`. I have been more successful with this setup so far and was wondering why I got a `Cannot get dependency key` **error** which lead me to #1414 and here. Among this error it brings some nice enhancements 👍

**Comment by Js-Brecht** [comment]
Rush is pretty great, but this issue was a problem for me, too.  I've been using a workaround for a while now, while this PR has been waiting.

Rush will download `rush-lib` to its cache in your home directory, and then it will continue to use that for as long as you are using the same rush version in your repository.  You can modify that lib to make it use the default store location.  This works for me:

```diff
file: ~/.rush/node-v10.18.1/rush-5.17.2/node_modules/@microsoft/rush-lib/lib/logic/InstallManager.js

- 857 args.push('--store', this._rushConfiguration.pnpmStoreFolder);
+ 857 // args.push('--store', this._rushConfiguration.pnpmStoreFolder);
```

YMMV, depending on what OS and what version of node/rush you're using.  Can just do a search for that line, though :+1:

**Comment by iclanton** [comment]
This seems like a really useful feature. I'd like to revise the design a bit, though. The `global` and `local` values for `pnpmStore` make sense, but I'm not seeing a common use case for `path` that wouldn't be better handled by setting the `RUSH_PNPM_STORE_PATH` env variable.

Here's how I'd expect this feature to be used:

- A repo maintainer decides whether it makes sense for their devs to use a local store or a global store on their machines. I'd guess most repos would want to use the global store, but repos that are backed by private feeds that don't proxy the public npmJS feed would want to use the local store as to not potentially pollute the global store.
- Builds running on CI set the `RUSH_PNPM_STORE_PATH`, which can then be cached between build machines.

I'm not really seeing how specifying the store path relative to the repo root statically in `rush.json` would be used. I also think the `RUSH_PNPM_STORE_PATH` should be allowed regardless of whether `pnpmStore` is set to `local` or `global`. If I'm missing a case here, can you please outline it for me?

File: apps/rush-lib/src/api/EnvironmentConfiguration.ts [hunk 1]
```diff
@@ -124,6 +145,13 @@ export class EnvironmentConfiguration {
             break;
           }
 
+          case EnvironmentVariableNames.RUSH_PNPM_STORE_PATH: {
+            EnvironmentConfiguration._pnpmStorePathOverride = (value && !options.doNotNormalizePaths)
+              ? EnvironmentConfiguration._normalizeDeepestParentFolderPath(value) || value
+              : value;
+              break;
```
Comment (iclanton) [inline apps/rush-lib/src/api/EnvironmentConfiguration.ts]:
```suggestion
            break;
```

File: apps/rush-lib/src/api/LastInstallFlag.ts [hunk 2]
```diff
@@ -4,6 +4,141 @@ import { FileSystem, JsonFile, JsonObject } from '@microsoft/node-core-library';
 
 export const LAST_INSTALL_FLAG_FILE_NAME: string = 'last-install.flag';
 
+/** @internal */
```
Comment (iclanton) [inline apps/rush-lib/src/api/LastInstallFlag.ts]: `@internal` is only needed if this is exported from the project's `index.ts`

File: apps/rush-lib/src/api/LastInstallFlag.ts [hunk 3]
```diff
@@ -4,6 +4,141 @@ import { FileSystem, JsonFile, JsonObject } from '@microsoft/node-core-library';
...
+   * If `deps` are defined for a property, all of those properties must match
+   * between old & new state in order for the error message to be generated.
+   */
+  private static _errors: IPropertyErrors = {
+    storePath: {
+      error: "Current PNPM store path does not match the last one used.  This may cause inconsistency in your builds.\n\n" +
+      "If you wish to install with the new store path, please run \"rush update --purge\"",
+      deps: ["packageManager"]
+    }
+  }
```
Comment (iclanton) [inline apps/rush-lib/src/api/LastInstallFlag.ts]: I think it would be simpler if we just stuck these inline in the `switch` statement in the constructor.

File: apps/rush-lib/src/api/RushConfiguration.ts [hunk 4]
```diff
@@ -151,6 +153,14 @@ export interface ICurrentVariantJson {
   variant: string | null; // Use `null` instead of `undefined` because `undefined` is not handled by JSON.
 }
 
+/**
+ * The paths used when resolving `pnpmStorePath`
+ */
+interface IPnpmOptionsRootPaths {
+  commonTempFolder: string;
+  projectRoot: string;
```
Comment (iclanton) [inline apps/rush-lib/src/api/RushConfiguration.ts]: We generally use the term `project` to refer to the individual projects inside the repo. I'd call this `repoRoot`

**Comment by Js-Brecht** [comment]
> I'm not really seeing how specifying the store path relative to the repo root statically in `rush.json` would be used.

You have a point.  It does the same thing as `RUSH_PNPM_STORE_PATH`, but represents a very specific opinion about where the store should be, wherever the repository is located.  I think I just included it because that's where my mind was at the time.  It makes perfect sense to rely solely on `RUSH_PNPM_STORE_PATH` for defining where that path should be located, because it really should be based on the environment.

> I also think the `RUSH_PNPM_STORE_PATH` should be allowed regardless of whether `pnpmStore` is set to `local` or `global`.

Also makes sense to me.  If that variable is defined in the environment, then it should definitely take precedence over the preference defined in the repository.  That would definitely make a repository more portable... wouldn't have to change the setting to `global` in order to define your own store location, then set it back to `local` when you want to commit.

Thanks, @iclanton.  I'll make some revisions later.

File: apps/rush-lib/src/logic/InstallManager.ts [hunk 5]
```diff
@@ -979,12 +986,15 @@ export class InstallManager {
...
+            if (
+              this._rushConfiguration.packageManager === 'pnpm' &&
+              this._rushConfiguration.pnpmOptions.pnpmStore === 'local'
+            ) {
               // If the installation has failed even after the retries, then pnpm store may
               // have got into a corrupted, irrecoverable state. Delete the store so that a
               // future install can create the store afresh.
               console.log(colors.yellow(`Deleting the "pnpm-store" folder`));
-              this._commonTempFolderRecycler.moveFolder(this._rushConfiguration.pnpmStoreFolder);
+              this._commonTempFolderRecycler.moveFolder(this._rushConfiguration.pnpmOptions.pnpmStorePath);
```
Comment (iclanton) [inline apps/rush-lib/src/logic/InstallManager.ts]: `rush purge --unsafe` should probably delete the global store. Is that case handled?

File: apps/rush-lib/src/api/LastInstallFlag.ts (hunk 2 above)
Comment (Js-Brecht) [reply apps/rush-lib/src/api/LastInstallFlag.ts]: I see.  I've gone ahead and removed the extraneous `@internal` markers

File: apps/rush-lib/src/api/LastInstallFlag.ts (hunk 3 above)
Comment (Js-Brecht) [reply apps/rush-lib/src/api/LastInstallFlag.ts]: I simplified the whole process for generating errors for the `last-install.flag` state.  Basically ripped out all of the abstraction, so it's much more straightforward now.

File: apps/rush-lib/src/api/RushConfiguration.ts (hunk 4 above)
Comment (Js-Brecht) [reply apps/rush-lib/src/api/RushConfiguration.ts]: Done.

File: apps/rush-lib/src/logic/InstallManager.ts (hunk 5 above)
Comment (Js-Brecht) [reply apps/rush-lib/src/logic/InstallManager.ts]: Seems strange to me, considering I would consider that PNPM territory, not Rush; but I'm not opposed.  I can see how it would make sense under certain circumstances, and it is, after all "unsafe" :laughing:.

There's still extra work I need to do for determining the global store location in as many cases as possible (i.e. if it is user configured by pnpm configs, or if it is changed in `.npmrc`), in order to address https://github.com/microsoft/rushstack/pull/1657#issuecomment-594824405

Please let me know if you see anything erroneous in the `PurgeManager.purgeUnsafe()` implementation.  It's not 100% ready to go, yet.  I need to do some testing, but I've run out of time tonight.

File: apps/rush-lib/src/api/LastInstallFlag.ts [hunk 6]
```diff
@@ -1,9 +1,54 @@
 import * as path from 'path';
 import * as _ from 'lodash';
 import { FileSystem, JsonFile, JsonObject } from '@microsoft/node-core-library';
+import { PackageManagerName } from './packageManager/PackageManager'
```
Comment (iclanton) [inline apps/rush-lib/src/api/LastInstallFlag.ts]:
```suggestion

import { PackageManagerName } from './packageManager/PackageManager'
```

File: apps/rush-lib/src/logic/InstallManager.ts [hunk 7]
```diff
@@ -1164,7 +1175,15 @@ export class InstallManager {
...
-      args.push('--store', this._rushConfiguration.pnpmStoreFolder);
+      // Only explicitly define the store path if `pnpmStore` is using the default, or has been set to
+      // 'local'.  If `pnpmStore` = 'global', then allow PNPM to use the system's default
+      // path.  In all cases, this will be overridden by RUSH_PNPM_STORE_PATH
+      if (
+        this._rushConfiguration.pnpmOptions.pnpmStore === 'local' ||
+        EnvironmentConfiguration.pnpmStorePathOverride
+      )
+          args.push('--store', this._rushConfiguration.pnpmOptions.pnpmStorePath);
+      }
```
Comment (iclanton) [inline apps/rush-lib/src/logic/InstallManager.ts]:
```suggestion
      ) {
        args.push('--store', this._rushConfiguration.pnpmOptions.pnpmStorePath);
      }
```

File: common/changes/@microsoft/rush/pnpm-store-options_2019-12-11-21-38.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Implement `pnpmOptions.pnpmStore` and RUSH_PNPM_STORE_PATH, to allow the end-user to define where PNPM will load its store.",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/pnpm-store-options_2019-12-11-21-38.json]:
```suggestion
      "comment": "Implement `pnpmOptions.pnpmStore` and RUSH_PNPM_STORE_PATH, to allow the end-user to define where PNPM will place its store.",
```

File: common/reviews/api/rush-lib.api.md [hunk 9]
```diff
@@ -230,9 +231,13 @@ export type PackageManagerName = 'pnpm' | 'npm' | 'yarn';
 // @public
 export class PnpmOptionsConfiguration {
     // Warning: (ae-forgotten-export) The symbol "IPnpmOptionsJson" needs to be exported by the entry point index.d.ts
-    // 
+    // Warning: (ae-forgotten-export) The symbol "IPnpmOptionsRootPaths" needs to be exported by the entry point index.d.ts
+    //
     // @internal
-    constructor(json: IPnpmOptionsJson);
+    constructor(json: IPnpmOptionsJson, rootPaths: IPnpmOptionsRootPaths);
+    // Warning: (ae-forgotten-export) The symbol "StoreOptions" needs to be exported by the entry point index.d.ts
```
Comment (iclanton) [inline common/reviews/api/rush-lib.api.md]: Make sure you export this from `src/index.ts`

File: apps/rush-lib/src/logic/InstallManager.ts (hunk 7 above)
Comment (Js-Brecht) [reply apps/rush-lib/src/logic/InstallManager.ts]: Thank you for catching that!  I don't know how I missed it

File: common/reviews/api/rush-lib.api.md (hunk 9 above)
Comment (Js-Brecht) [reply common/reviews/api/rush-lib.api.md]: Resolving this now, but I have one question:

If I don't mark this interface as `@public` or `@internal` (or the other various tags), it will complain.  AFAICT, it's only an interface that will be used internally, but if I mark it `@internal`, it complains about it not having an underscore.

Is this because the `PnpmOptionsConfiguration` class is marked `@public`?  Is it an issue for the interfaces to be `@public` as well?

Comment (iclanton) [reply common/reviews/api/rush-lib.api.md]: Yeah it's complaining because `PnpmOptionsConfiguration` is `@public` and `pnpmStore`'s type is `StoreOptions` and `pnpmStore` is a public member of `PnpmOptionsConfiguration`.

Just a thought on this from the perspective of the API review - can we rename the type to `PnpmStoreOptions`?

Comment (Js-Brecht) [reply common/reviews/api/rush-lib.api.md]: I actually did that in this last pass; I just haven't committed it yet, because I was waiting for your feedback.

The main issue is with `IPnpmOptionsJson` and `IPnpmOptionsRootPaths`.  `IPnpmStoreOptions` is marked `@public`, so it's not complaining about that one.  I just had to export it from `index.ts`.

Comment (Js-Brecht) [reply common/reviews/api/rush-lib.api.md]: I went ahead and marked the interfaces public... they're just types, after all.  Please let me know if that's an issue.  Also added some docs to describe the interface members.  Trying to iron out the these issue in the API docs.

File: common/reviews/api/rush-lib.api.md [hunk 10]
```diff
@@ -144,6 +145,22 @@ export class IndividualVersionPolicy extends VersionPolicy {
     validate(versionString: string, packageName: string): void;
 }
 
+// @public
+export interface IPnpmOptionsJson {
```
Comment (iclanton) [inline common/reviews/api/rush-lib.api.md]: This should be marked as `@internal`

File: common/reviews/api/rush-lib.api.md [hunk 11]
```diff
@@ -144,6 +145,22 @@ export class IndividualVersionPolicy extends VersionPolicy {
...
+    pnpmStore?: IPnpmStoreOptions;
+    resolutionStrategy?: ResolutionStrategy;
+    strictPeerDependencies?: boolean;
+}
+
+// @public
+export interface IPnpmOptionsRootPaths {
```
Comment (iclanton) [inline common/reviews/api/rush-lib.api.md]: This one should also be `@internal`

File: common/reviews/api/rush-lib.api.md [hunk 12]
```diff
@@ -144,6 +145,22 @@ export class IndividualVersionPolicy extends VersionPolicy {
...
+    commonTempFolder: string;
+    repoRoot: string;
+}
+
+// @public
+export type IPnpmStoreOptions = 'local' | 'global';
```
Comment (iclanton) [inline common/reviews/api/rush-lib.api.md]: This isn't an interface, so its name shouldn't start with an `I`

File: apps/rush-lib/src/index.ts [hunk 13]
```diff
@@ -15,6 +15,9 @@ export {
   ITryFindRushJsonLocationOptions,
   ResolutionStrategy,
   PnpmOptionsConfiguration,
+  IPnpmOptionsJson,
+  IPnpmStoreOptions,
+  IPnpmOptionsRootPaths,
```
Comment (iclanton) [inline apps/rush-lib/src/index.ts]:
```suggestion
  IPnpmOptionsJson as _IPnpmOptionsJson,
  IPnpmStoreOptions,
  IPnpmOptionsRootPaths as _IPnpmOptionsRootPaths,
```

Comment (Js-Brecht) [reply apps/rush-lib/src/index.ts]: Oh, I see.  Thanks!

File: common/reviews/api/rush-lib.api.md (hunk 10 above)
Comment (Js-Brecht) [reply common/reviews/api/rush-lib.api.md]: Fixed

File: common/reviews/api/rush-lib.api.md (hunk 11 above)
Comment (Js-Brecht) [reply common/reviews/api/rush-lib.api.md]: Fixed

File: common/reviews/api/rush-lib.api.md (hunk 12 above)
Comment (Js-Brecht) [reply common/reviews/api/rush-lib.api.md]: Fixed

File: apps/rush-lib/src/api/LastInstallFlag.ts [hunk 14]
```diff
@@ -27,17 +73,56 @@ export class LastInstallFlag {
...
+   */
+  private _stateHasError(oldState: JsonObject): LastInstallFlagError | void {
+    const pkgManager: PackageManagerName = this._state.packageManager;
+    switch (pkgManager) {
+      case 'pnpm': {
+        if (
+          ( // Only throw an error if the package manager hasn't changed from pnpm
+            oldState.packageManager === pkgManager
+          ) && ( // Throw if the store path changed
+            oldState.storePath &&
```
Comment (Js-Brecht) [inline apps/rush-lib/src/api/LastInstallFlag.ts]: @iclanton I just realized that this should be `oldState.storePath !== undefined`.

If the default global path is used, the value here will be "".  So when the store value is changed to a different location, this condition will not catch the error state, and it will fall through to PNPM, which will throw an exception instead.

Additionally, this check has to be here, because if somebody is changing package managers, then this property will be `undefined`, which is **not** an error state.

This is something the tests should have caught, but I don't think I checked for "" -> "/some/path"... just "/some/path" !== "/some/other/path" :man_facepalming:
